Cloudbuild plugin

ebarriosjr opened the pull request:
## Hey, I just made a Pull Request!

Added cloudbuild CI/CD to backstage.

<!-- Please describe what you added, and add a screenshot if possible.
     That makes it easier to understand the change so we can :shipit: faster. -->

#### :heavy_check_mark: Checklist

<!--- Put an `x` in all the boxes that apply: -->

- [ ] All tests are passing `yarn test`
- [ ] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [ ] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes

ebarriosjr commented:
Main Page:
![main-cloudbuild](https://user-images.githubusercontent.com/7311854/93585434-1a0aee00-f9a7-11ea-9abf-69adf8b86705.png)

Details Page:
![details-cloudbuild](https://user-images.githubusercontent.com/7311854/93585532-3eff6100-f9a7-11ea-8301-ce516ab0659a.png)

stefanalund commented:
👏

stefanalund commented on plugins/cloudbuild/README.md:
```diff
@@ -0,0 +1,13 @@
+# cloudbuild
+
+Welcome to the cloudbuild plugin!
```
This is Google Cloud Build, correct? I think that should be a bit more clear.

ebarriosjr replied:
I changed it to Google cloudbuild

freben commented on plugins/cloudbuild/src/components/useProjectName.ts:
```diff
@@ -0,0 +1,27 @@
+import { useAsync } from 'react-use';
+import { Entity } from '@backstage/catalog-model';
+
+export const CLOUDBUILD_ANNOTATION = 'google.com/cloudbuild';
```
a bit unclear what the format of this one is - either we could call it `google.com/cloudbuild-project-slug` similar to the other [well known annotations](https://backstage.io/docs/features/software-catalog/well-known-annotations) if it's just a project, or a generic `google.com/project-slug` if it is commonly the case that all features are keyed off of the same thing.

freben commented on plugins/cloudbuild/src/components/useWorkflowRuns.ts:
```diff
@@ -0,0 +1,97 @@
+    const token = await auth.getAccessToken([
+      'https://www.googleapis.com/auth/cloud-platform',
+    ]);
+    return api
```
just noting that you could `await` this too to avoid the extra nesting of the `then` below :)

freben commented on plugins/cloudbuild/src/components/useWorkflowRuns.ts:
```diff
@@ -0,0 +1,97 @@
+        (
+          workflowRunsData: ActionsListWorkflowRunsForRepoResponseData,
+        ): WorkflowRun[] => {
+          setTotal(workflowRunsData.total_count);
```
yeah at this point i think you should just do `setTotal(workflowRunsData.builds.length)` and skip that extra field...

freben commented on plugins/cloudbuild/src/components/useWorkflowRuns.ts:
```diff
@@ -0,0 +1,97 @@
+          // Transformation here
+          return workflowRunsData.builds.map(run => ({
+            message: run.substitutions.REPO_NAME,
+            id: `${run.id}`,
```
just `id: run.id` right?

freben commented on plugins/cloudbuild/src/components/useWorkflowRuns.ts:
```diff
@@ -0,0 +1,97 @@
+          return workflowRunsData.builds.map(run => ({
+            message: run.substitutions.REPO_NAME,
+            id: `${run.id}`,
+            onReRunClick: async () => {
```
just `rerun()` right, we don't care it's a click here

or move this code to the caller instead and keep this a raw data object

ebarriosjr commented:
Hi @freben I think I addressed all the comments you left for me. Thank you very much for your feedback. Please let me know if there is anything else I should change

ebarriosjr commented on plugins/cloudbuild/src/api/CloudbuildClient.ts:
```diff
@@ -0,0 +1,115 @@
+
+export class CloudbuildClient implements CloudbuildApi {
+  async reRunWorkflow({
+    token,
```
This is amazing!

freben commented on plugins/cloudbuild/package.json:
```diff
@@ -0,0 +1,56 @@
+  "dependencies": {
+    "@backstage/catalog-model": "^0.1.1-alpha.23",
+    "@backstage/core": "^0.1.1-alpha.23",
+    "@backstage/core-api": "^0.1.1-alpha.23",
```
you shouldn't depend on core-api; what you need should exist exported from core

freben commented on plugins/cloudbuild/src/api/CloudbuildApi.ts:
```diff
@@ -0,0 +1,54 @@
+    token: string;
+    projectId: string;
+    id: string;
+  }) => Promise<ActionsGetWorkflowResponseData>;
```
the same comment as above applies to all methods in here - you want to remove the token, and you don't need to deconstruct the argument (repeating the fields)

freben commented on plugins/cloudbuild/src/components/Cards/Cards.tsx:
```diff
@@ -0,0 +1,120 @@
+  branch: string;
+}) => {
+  const classes = useStyles();
+  if (error) return <Typography>Couldn't fetch latest {branch} run</Typography>;
```
There's a `WarningPanel` component in core that's perfect for error feedback to users

freben commented on plugins/cloudbuild/src/components/Cards/Cards.tsx:
```diff
@@ -0,0 +1,120 @@
+  return (
+    <InfoCard title={`Last ${branch} build`}>
+      <WidgetContent
+        error={error}
```
oh hey i see now that you use both the errorApi AND a text output for the same error - maybe best to decide which one of them to use - could be reasonable to inside the InfoCard have a switch that if there's an error draw a WarningPanel, if it's loading draw a progress bar, otherwise draw WidgetContent.

That way, WidgetContent does not have to have neither know of loading nor error states. And then you can remove the errorApi usage here

ebarriosjr replied:
This I copied from the github-actions plugin. You can see it here: https://github.com/spotify/backstage/blob/17c30eaf3efd026032110c8d80217e63f4385b9f/plugins/github-actions/src/components/Cards/Cards.tsx#L104 
I used this plugin as a base for writing the google Cloud Build plugin.

freben commented on plugins/cloudbuild/src/components/WorkflowRunDetails/WorkflowRunDetails.tsx:
```diff
@@ -0,0 +1,148 @@
+  } else if (loading) {
+    return <LinearProgress />;
+  } else if (details.value?.logUrl === undefined) {
+    return <LinearProgress />;
```
this is peculiar - does data arrive twice, once without this field and once with it?

ebarriosjr replied:
Yes, it happens that sometimes i get an undefined on the first request but on the next one it is defined. I am not entirely sure why... any help with this would be appreciated.

ebarriosjr commented on plugins/cloudbuild/README.md:
```diff
@@ -0,0 +1,13 @@
+# Google cloudbuild
```
Renamed it as requested

ebarriosjr commented on plugins/cloudbuild/src/api/CloudbuildClient.ts:
```diff
@@ -0,0 +1,115 @@
+    runId: string;
+  }): Promise<any> {
+    return await fetch(
+      `https://cloudbuild.googleapis.com/v1/projects/${projectId}/builds/${runId}:retry`,
```
added `encodeURIComponent` as requested

ebarriosjr commented on plugins/cloudbuild/src/api/CloudbuildClient.ts:
```diff
@@ -0,0 +1,115 @@
+    token: string;
+    projectId: string;
+    runId: string;
+  }): Promise<any> {
```
Change it to `Promise<void>` as requested.

benjdlambert commented:
@freben could you re-review this?

ebarriosjr commented:
Hi @freben, shoudl I resolve all the previous comments? I see that every time I do this a message is generated on the bots channel and i don't want to spam it... :P

benjdlambert commented:
@ebarriosjr could you fix the merge conflicts also? :) It's on his list to-do, so he should get to it soon 👍

freben reviewed (approved):
Let's get this in! Gotta fix the conflict is all

ebarriosjr commented:
@freben all conflicts have been fixed.
